pinecone/index_connection.go

```diff
@@ -0,0 +1,358 @@
+	}
+
+	vectors := make(map[string]*Vector, len(res.Vectors))
+	for id, vector := range res.Vectors {
+		vectors[id] = toVector(vector)
+	}
+
+	return &FetchVectorsResponse{
+		Vectors:   vectors,
+		Namespace: res.Namespace,
+		Usage:     toUsage(res.Usage),
+	}, nil
+}
+
+type ListVectorsRequest struct {
+	Prefix          *string
+	Limit           *uint32
+	PaginationToken *string
+	Namespace       string
+}
+
+type ListVectorsResponse struct {
+	VectorIds           []*string
+	Namespace           string
+	Usage               *Usage
+	NextPaginationToken *string
+}
+
+func (idx *IndexConnection) ListVectors(ctx *context.Context, in *ListVectorsRequest) (*ListVectorsResponse, error) {
+	req := &data.ListRequest{
+		Prefix:          in.Prefix,
+		Limit:           in.Limit,
+		PaginationToken: in.PaginationToken,
+		Namespace:       in.Namespace,
+	}
+	res, err := (*idx.dataClient).List(idx.akCtx(*ctx), req)
+	if err != nil {
+		return nil, err
+	}
+
+	vectorIds := make([]*string, len(res.Vectors))
+	for i := 0; i < len(res.Vectors); i++ {
+		vectorIds[i] = &res.Vectors[i].Id
+	}
+
+	return &ListVectorsResponse{
+		VectorIds:           vectorIds,
+		Namespace:           res.Namespace,
+		Usage:               &Usage{ReadUnits: res.Usage.ReadUnits},
+		NextPaginationToken: toPaginationToken(res.Pagination),
+	}, nil
+}
+
+type QueryByVectorRequest struct {
+	Vector          []float32
+	Namespace       string
+	TopK            uint32
+	Filter          *Filter
+	IncludeValues   bool
+	IncludeMetadata bool
+	SparseValues    *SparseValues
+}
+
+type QueryVectorsResponse struct {
+	Matches   []*ScoredVector
+	Namespace string
+	Usage     *Usage
+}
+
+func (idx *IndexConnection) QueryByVector(ctx *context.Context, in *QueryByVectorRequest) (*QueryVectorsResponse, error) {
+	req := &data.QueryRequest{
+		Namespace:       in.Namespace,
+		TopK:            in.TopK,
+		Filter:          in.Filter,
+		IncludeValues:   in.IncludeValues,
+		IncludeMetadata: in.IncludeMetadata,
+		Vector:          in.Vector,
+		SparseVector:    sparseValToGrpc(in.SparseValues),
+	}
+
+	return idx.query(ctx, req)
+}
+
+type QueryByIdRequest struct {
+	Id              string
+	Namespace       string
+	TopK            uint32
+	Filter          *Filter
+	IncludeValues   bool
+	IncludeMetadata bool
+	SparseValues    *SparseValues
+}
+
+func (idx *IndexConnection) QueryById(ctx *context.Context, in *QueryByIdRequest) (*QueryVectorsResponse, error) {
+	req := &data.QueryRequest{
+		Id:              in.Id,
+		Namespace:       in.Namespace,
+		TopK:            in.TopK,
+		Filter:          in.Filter,
+		IncludeValues:   in.IncludeValues,
+		IncludeMetadata: in.IncludeMetadata,
+		SparseVector:    sparseValToGrpc(in.SparseValues),
+	}
+
+	return idx.query(ctx, req)
+}
+
+type DeleteVectorsRequest struct {
+	Namespace string
+	Ids       []string
+	Filter    *Filter
+	DeleteAll bool
+}
+
+func (idx *IndexConnection) DeleteVectors(ctx *context.Context, in *DeleteVectorsRequest) error {
```

Comment:
I gave similar feedback recently on a java PR, but I think we should consider breaking out some different helper methods for things like `DeleteById`, `DeleteByFilter`, and `DeleteByNamespace` (e.g. DeleteAll) to guide the user better, as the various options accepted by our API are mutually exclusive.

Reply:
1f9a12d